csharp/ql/src/semmle/code/csharp/dataflow/SignAnalysis.qll

```diff
@@ -0,0 +1,595 @@
+/**
+ * Provides sign analysis to determine whether expression are always positive
+ * or negative.
+ *
+ * The analysis is implemented as an abstract interpretation over the
+ * three-valued domain `{negative, zero, positive}`.
+ */
+
+private import csharp
+private import Ssa
+private import SsaReadPosition
+private import ConstantUtils
+private import RangeUtils
+private import semmle.code.csharp.controlflow.Guards as G
+private import Linq.Helpers as Linq
+
+private class Guard = G::Guard;
+
+private newtype TSign =
+  TNeg() or
+  TZero() or
+  TPos()
+
+private class Sign extends TSign {
+  string toString() {
+    result = "-" and this = TNeg()
+    or
+    result = "0" and this = TZero()
+    or
+    result = "+" and this = TPos()
+  }
+
+  Sign inc() {
+    this = TNeg() and result = TNeg()
+    or
+    this = TNeg() and result = TZero()
+    or
+    this = TZero() and result = TPos()
+    or
+    this = TPos() and result = TPos()
+  }
+
+  Sign dec() { result.inc() = this }
+
+  Sign neg() {
+    this = TNeg() and result = TPos()
```

Comment:
This is probably an edge case:
Consider a 16-bit integer, then `abs(-32768) = -32768` should be true, which is still negative.
(Unless C# throws an exception in this case?)

Reply:
Yes, I don't think we want to take this into account. Similarly, overflow is not considered.